collect_app/src/main/java/org/odk/collect/android/formentry/ODKView.java

```diff
@@ -405,48 +405,49 @@ private void addIntentLaunchButton(Context context, FormEntryPrompt[] questionPr
         launchIntentButton.setText(buttonText);
         launchIntentButton.setTextSize(TypedValue.COMPLEX_UNIT_DIP, QuestionFontSizeUtils.getQuestionFontSize() + 2);
         launchIntentButton.setVisibility(VISIBLE);
-        launchIntentButton.setOnClickListener(new OnClickListener() {
-            @Override
-            public void onClick(View v) {
-                String intentName = ExternalAppsUtils.extractIntentName(intentString);
-                Map<String, String> parameters = ExternalAppsUtils.extractParameters(
-                        intentString);
-
-                Intent i = new Intent(intentName);
-                try {
-                    ExternalAppsUtils.populateParameters(i, parameters,
-                            c.getIndex().getReference());
-
-                    for (FormEntryPrompt p : questionPrompts) {
-                        IFormElement formElement = p.getFormElement();
-                        if (formElement instanceof QuestionDef) {
-                            TreeReference reference =
-                                    (TreeReference) formElement.getBind().getReference();
-                            IAnswerData answerValue = p.getAnswerValue();
-                            Object value =
-                                    answerValue == null ? null : answerValue.getValue();
-                            switch (p.getDataType()) {
-                                case Constants.DATATYPE_TEXT:
-                                case Constants.DATATYPE_INTEGER:
-                                case Constants.DATATYPE_DECIMAL:
-                                case Constants.DATATYPE_BINARY:
-                                    i.putExtra(reference.getNameLast(),
-                                            (Serializable) value);
-                                    break;
-                            }
+        launchIntentButton.setOnClickListener(view -> {
+            String intentName = ExternalAppsUtils.extractIntentName(intentString);
+            Map<String, String> parameters = ExternalAppsUtils.extractParameters(intentString);
+
+            Intent i = new Intent(intentName);
+            if (i.resolveActivity(Collect.getInstance().getPackageManager()) == null) {
+                i = Collect.getInstance().getPackageManager().getLaunchIntentForPackage(intentName);
+                i.setFlags(0);
```

Comment:
I can't remember the details of why this is needed but can try to pull it up again if needed. There are flags set by default when doing `getLaunchIntentForPackage` and they suppress something, maybe returning a result.

Reply:
I tested the fix without the flag (not here but using external widgets) and it seems to work so would be good if you could double check it.

Reply:
Now I remember that at some point I got into a state where the OpenRDT app was immediately returning its result rather than first going through its UI flow and then returning. That led me to [this Stackoverflow post](https://stackoverflow.com/a/24312398/137744). The OpenRDT examples I was sent (not public, unfortunately) do use `setFlags(0)`.

>  getLaunchIntentForPackage("com.your.package.here") would create an intent with Intent.FLAG_ACTIVITY_NEW_TASK added by default, so call: in.setFlags(0);

Always creating the target activity seems like the way to go to me. There are differences in Android versions with how the other launch modes work and their behavior will also depend on whether or not an activity is already in memory.

Reply:
Is it just me or are the docs around this particularly bad?

Reply:
Ok this is confusing but makes more sense now thanks!